// ngx-table-generator/esm5/lib/utils.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var TableUtils = /** @class */ (function () {
    function TableUtils() {
    }
    /**
     * @param {?} array
     * @param {?} value
     * @return {?}
     */
    TableUtils.prototype.deleteInstance = /**
     * @param {?} array
     * @param {?} value
     * @return {?}
     */
    function (array, value) {
        for (var i = 0; i < array.length; i++) {
            if (value === array[i]) {
                array.splice(i, 1);
                break;
            }
        }
        return array;
    };
    /**
     * @param {?} data
     * @param {?} arrayData
     * @param {?} id
     * @return {?}
     */
    TableUtils.prototype.chargeMultiCheckArray = /**
     * @param {?} data
     * @param {?} arrayData
     * @param {?} id
     * @return {?}
     */
    function (data, arrayData, id) {
        /** @type {?} */
        var array = [];
        var _loop_1 = function (i) {
            arrayData.map((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item === data[i][id]) {
                    array[i] = true;
                }
            }));
        };
        for (var i = 0; i < data.length; i++) {
            _loop_1(i);
        }
        return array;
    };
    /**
     * @param {?} val
     * @param {?} decimals
     * @return {?}
     */
    TableUtils.prototype.transformToNumber = /**
     * @param {?} val
     * @param {?} decimals
     * @return {?}
     */
    function (val, decimals) {
        if (val !== undefined && val !== null) {
            decimals = decimals || decimals === 0 ? decimals : 2;
            if (isNaN(val)) {
                return val;
            }
            else {
                val = val.toLocaleString('es-ES', { minimumFractionDigits: decimals, maximumFractionDigits: decimals });
                if (val.includes(',')) {
                    val = this.covertMillarWithDecimalToSpanish(val);
                }
                else {
                    val = this.convertMillarToSpanish(val);
                }
                return val;
            }
        }
        else {
            return '';
        }
    };
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    TableUtils.prototype.convertMillarToSpanish = /**
     * @private
     * @param {?} val
     * @return {?}
     */
    function (val) {
        /** @type {?} */
        var splitVal = val.split('');
        if (splitVal.length === 4) {
            splitVal.splice(1, 0, '.');
        }
        val = '';
        splitVal.forEach((/**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            val += element;
        }));
        return val;
    };
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    TableUtils.prototype.covertMillarWithDecimalToSpanish = /**
     * @private
     * @param {?} val
     * @return {?}
     */
    function (val) {
        /** @type {?} */
        var splitVal = val.split(',');
        if (splitVal[0].length === 4) {
            /** @type {?} */
            var naturalNum = splitVal[0];
            splitVal[0] = '';
            naturalNum = naturalNum.split('');
            naturalNum.splice(1, 0, '.');
            naturalNum.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                splitVal[0] += element;
            }));
            val = splitVal[0] + "," + splitVal[1];
        }
        return val;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    TableUtils.prototype.copyArray = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var _this = this;
        /** @type {?} */
        var result = null;
        if (!item) {
            return result;
        }
        if (Array.isArray(item)) {
            result = [];
            item.forEach((/**
             * @param {?} element
             * @return {?}
             */
            function (element) {
                result.push(_this.copyArray(element));
            }));
        }
        else if (item instanceof Object && !(item instanceof Function)) {
            result = {};
            for (var key in item) {
                if (key) {
                    result[key] = this.copyArray(item[key]);
                }
            }
        }
        return result || item;
    };
    /**
     * @param {?} conditionalMulticheck
     * @param {?} keysMulticheck
     * @param {?} valueMulticheckToCompare
     * @param {?} data
     * @return {?}
     */
    TableUtils.prototype.chargeShowMulticheck = /**
     * @param {?} conditionalMulticheck
     * @param {?} keysMulticheck
     * @param {?} valueMulticheckToCompare
     * @param {?} data
     * @return {?}
     */
    function (conditionalMulticheck, keysMulticheck, valueMulticheckToCompare, data) {
        var _this = this;
        /** @type {?} */
        var showArray = [];
        data.forEach((/**
         * @param {?} element
         * @param {?} index
         * @return {?}
         */
        function (element, index) {
            showArray[index] = true;
        }));
        if (conditionalMulticheck && keysMulticheck && valueMulticheckToCompare) {
            var _loop_2 = function (i) {
                data.forEach((/**
                 * @param {?} element
                 * @param {?} index
                 * @return {?}
                 */
                function (element, index) {
                    if (!_this.suboperationConditional(conditionalMulticheck[i], valueMulticheckToCompare[i], element[keysMulticheck[i]])
                        && showArray[index]) {
                        showArray[index] = false;
                    }
                }));
            };
            for (var i = 0; i < keysMulticheck.length; i++) {
                _loop_2(i);
            }
        }
        return showArray;
    };
    /**
     * @param {?} conditional
     * @param {?} valueToCompare
     * @param {?} value
     * @return {?}
     */
    TableUtils.prototype.suboperationConditional = /**
     * @param {?} conditional
     * @param {?} valueToCompare
     * @param {?} value
     * @return {?}
     */
    function (conditional, valueToCompare, value) {
        switch (conditional) {
            case '<':
                return parseFloat(value) < parseFloat(valueToCompare);
            case '>':
                return parseFloat(value) > parseFloat(valueToCompare);
            case '===':
                return (value || value === 0 || value === false ? value.toString() : value) ===
                    (valueToCompare || valueToCompare === 0 || valueToCompare === false ? valueToCompare.toString() : valueToCompare);
            case '>=':
                return value >= valueToCompare;
            case '<=':
                return value <= valueToCompare;
            case '<day':
                return (new Date(value).setHours(0) - new Date().setHours(0, 0, 0, 0)) < 0;
            case '>day':
                return (new Date(value).setHours(0) - new Date().setHours(0, 0, 0, 0)) > 0;
        }
    };
    TableUtils.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ TableUtils.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TableUtils_Factory() { return new TableUtils(); }, token: TableUtils, providedIn: "root" });
    return TableUtils;
}());
export { TableUtils };
